controller/python: add device attestation revocation support

shubhamdp opened the pull request:
- Added the required changes in the python binding.
- Added the cli option to matter testing framework for setting the dac revocation set path.
- src/credentials: add an API to DeviceAttestationVerifier interface to set the revocation delegate

#### Testing

- Manually tested by running TC_RR_1_1 and verified that commissioning stops with error "dac-revoked" in first case. And Test pass successfully when used without the arguments.

```
# Terminal 1
./out/host/chip-lighting-app --dac_provider credentials/test/revoked-attestation-certificates/dac-provider-test-vectors/revoked-pai.json

# Terminal 2
./scripts/build_python.sh --enable_ble true -i out/py-env

python3 src/python_testing/TC_RR_1_1.py -m on-network -p 20202021 -d 3840 --int-arg use_pase_only:0 --dac-revocation-set-path credentials/test/revoked-attestation-certificates/revocation-sets/revocation-set-for-paa.json
```

NOTE: Used the `dac-provider-test-vectors` from #37122

tehampson commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -2018,7 +2018,7 @@ class ChipDeviceController(ChipDeviceControllerBase):
     def __init__(self, opCredsContext: ctypes.c_void_p, fabricId: int, nodeId: int, adminVendorId: int, catTags: typing.List[int] = [
-    ], paaTrustStorePath: str = "", useTestCommissioner: bool = False, fabricAdmin: typing.Optional[FabricAdmin.FabricAdmin] = None, name: str = '', keypair: typing.Optional[p256keypair.P256Keypair] = None):
+    ], paaTrustStorePath: str = "", dacRevocationSetPath: str = "",  useTestCommissioner: bool = False, fabricAdmin: typing.Optional[FabricAdmin.FabricAdmin] = None, name: str = '', keypair: typing.Optional[p256keypair.P256Keypair] = None):
         super().__init__(
```
My approval is with the assumption:
* Ordering comment above it addressed 
* https://github.com/project-chip/connectedhomeip/pull/37134 lands first so people can replicate your test

shubhamdp replied:
Thanks @tehampson, #37134 is landed and I have addressed the reviews.

tehampson reviewed (dismissed):
Conditional approval

tcarmelveilleux commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -468,9 +478,9 @@ PyChipError pychip_OpCreds_AllocateControllerForPythonCommissioningFLow(
-                                              const char * paaTrustStorePath, bool useTestCommissioner,
-                                              bool enableServerInteractions, CASEAuthTag * caseAuthTags, uint32_t caseAuthTagLen,
-                                              chip::python::pychip_P256Keypair * operationalKey)
+                                              const char * paaTrustStorePath, const char * dacRevocationSetPath,
```
Please do not add this argument and let's configure the commissioner instance separately

shubhamdp replied:
Makes sense, added another API to configure the revocation set path.

tcarmelveilleux commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -492,7 +502,10 @@ PyChipError pychip_OpCreds_AllocateController(OpCredsContext * context, chip::Co
     // Initialize device attestation verifier
     const chip::Credentials::AttestationTrustStore * testingRootStore = GetTestFileAttestationTrustStore(paaTrustStorePath);
-    chip::Credentials::DeviceAttestationVerifier * dacVerifier        = chip::Credentials::GetDefaultDACVerifier(testingRootStore);
+    chip::Credentials::DeviceAttestationRevocationDelegate * dacRevocationDelegate =
+        GetAttestationRevocationDelegate(dacRevocationSetPath);
```
This is what should be moved to a new API in ChipDeviceController

tcarmelveilleux commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -492,7 +502,10 @@ PyChipError pychip_OpCreds_AllocateController(OpCredsContext * context, chip::Co
+    chip::Credentials::DeviceAttestationRevocationDelegate * dacRevocationDelegate =
+        GetAttestationRevocationDelegate(dacRevocationSetPath);
+    chip::Credentials::DeviceAttestationVerifier * dacVerifier =
+        chip::Credentials::GetDefaultDACVerifier(testingRootStore, dacRevocationDelegate);
```
Now I see that coupling the DAC revocation delegate to the default DAC verifier is what causes you to have to do this. It should still be possible to UPDATE the config of the DefaultDACVerifier, without needing to initially provide an instance of DAC revocation delegate.  Suggest passing nullptr here.

shubhamdp replied:
dynamic cast is failing so added an API which returns the pointer to the instance of type `DefaultDACVerifier`

shubhamdp replied:
Added a virtual function in `DeviceAttestationVerifier::TrySettingRevocationDelegate()` which returns `NOT_IMPLEMENTED` and has an implementation in `DefaultDACVerifier`.

tcarmelveilleux commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -2018,7 +2018,7 @@ class ChipDeviceController(ChipDeviceControllerBase):
 
     def __init__(self, opCredsContext: ctypes.c_void_p, fabricId: int, nodeId: int, adminVendorId: int, catTags: typing.List[int] = [
-    ], paaTrustStorePath: str = "", useTestCommissioner: bool = False, fabricAdmin: typing.Optional[FabricAdmin.FabricAdmin] = None, name: str = '', keypair: typing.Optional[p256keypair.P256Keypair] = None):
+    ], paaTrustStorePath: str = "", dacRevocationSetPath: str = "",  useTestCommissioner: bool = False, fabricAdmin: typing.Optional[FabricAdmin.FabricAdmin] = None, name: str = '', keypair: typing.Optional[p256keypair.P256Keypair] = None):
```
+1

shubhamdp replied:
Moved the argument at the end.

tehampson commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -700,4 +710,28 @@ PyChipError pychip_GetCompletionError()
+PyChipError pychip_DeviceController_SetDACRevocationSetPath(const char * dacRevocationSetPath)
+{
+    chip::Credentials::DeviceAttestationRevocationDelegate * dacRevocationDelegate =
+        GetTestAttestationRevocationDelegate(dacRevocationSetPath);
```
Doesn't this mean that you can only ever `dacRevocationSetPath` to be from the test attestation?

shubhamdp replied:
Yes, as far as this impl goes, similar to paa-trust-store-path we can set the dac-revocation-set-path. We have [test implementation](https://github.com/project-chip/connectedhomeip/blob/master/src/credentials/attestation_verifier/TestDACRevocationDelegateImpl.h) of the [DeviceAttestationRevocationDelegate](https://github.com/project-chip/connectedhomeip/blob/master/src/credentials/attestation_verifier/DeviceAttestationVerifier.h#L426) which uses the path based logic.

tehampson commented on src/controller/python/OpCredsBinding.cpp:
```diff
@@ -700,4 +710,17 @@ PyChipError pychip_GetCompletionError()
+        GetTestAttestationRevocationDelegate(dacRevocationSetPath);
+    VerifyOrReturnError(dacRevocationDelegate != nullptr, ToPyChipError(CHIP_ERROR_INVALID_ARGUMENT));
+
+    chip::Credentials::DefaultDACVerifier * dacVerifier = chip::Credentials::GetDefaultDACVerifierInstance();
```
I am not following why you cannot get `DefaultDACVerifier *` from `GetDefaultDACVerifier()`. Something seems off to me here. I know you mentioned that `dynamic_cast` didn't work, but my follow up question is why didn't it work. It really seems like we are putting up a bandaid to fix something that we should understand why the correct approach is not working. And if there is a legitimate reason why that isn't possible we need to communicate that in comment so the next person reviewing this section of code understand why this approach was taken

tehampson replied:
Looking at this a little more, would it make sense for `DeviceAttestationVerifier` to have a virtual function `CHIP_ERROR TrySetRevocationDelegate` that by default returns `CHIP_ERROR_NOT_IMPLEMENTED`, and in `DefaultDACVerifier` you override it to do what you want?

shubhamdp replied:
`dynamic_cast` is not possible because because of `-fno-rtti` flag is set when building the python controller. ( I see few failures in CI)
On adding that flag, we don't get the run time type information which stops us from performing the `dynamic_cast`. But, we get some space savings.

shubhamdp replied:
Sorry, that I did not mention the reason why `dynamic_cast` was failing, I should have.

shubhamdp replied:
I did gave a though about adding `SetRevocationDelegate()` into `DeviceAttestationVerifier`, but after looking into current members of [DeviceAttestationVerifier](https://github.com/project-chip/connectedhomeip/blob/master/src/credentials/attestation_verifier/DeviceAttestationVerifier.h#L263), I don't think its a right place to have add that API.

tehampson replied:
Why doesn't is seem like an okay place for it? Doesn't `DeviceAttestationVerifier` already have a pure virtual method for `CheckForRevokedDACChain`?

shubhamdp replied:
I think not adding a `SetRevocationDelegate()` there keeps that interface clean and `CheckForRevokedDACChain()` can be implemented without need of revocation delegate.

If we add `SetRevocationDelegate()` there, then we would be forcing the derived implementations to have that method and implement it somehow, even if they don't want to use that interface.

Adding a `Set_` method which does nothing may not be a good thing, IMO.

How I am seeing the `DefaultDACVerifier` is it implements methods in `DeviceAttestationVerifier` + introduces the test implementations of paa-trust-store and dac-revocation-set.

tehampson replied:
Sorry I am not following how something like the following is "forcing derived implementations to have a method", you are allowing implementations to optionally provide that ability:

```
virtual CHIP_ERROR TrySettingRevocationDelegate(DeviceAttestationRevocationDelegate * revocationDelegate)
{
    return CHIP_ERROR_NOT_IMPLEMENTED
}
```

shubhamdp replied:
Thanks @tehampson for clarifying that adding `TrySettingRevocationDelegate()` is better than the approach that I used earlier.

I have force pushed the branch, and re-wrote the history in order to make it cleaner.

Please re-review the changes.

shubhamdp committed:
src/credentials: add an API to DeviceAttestationVerifier interface to
set the revocation delegate

shubhamdp committed:
controller/python: add device attestation revocation support

- Added the cli option to matter testing framework for setting the dac
  revocation set path
- Added the required changes in the python bindings
- Added an API to the python controller to set the dac revocation set
  path.

arkq commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.h:
```diff
@@ -83,9 +83,10 @@ class DefaultDACVerifier : public DeviceAttestationVerifier
     CsaCdKeysTrustStore * GetCertificationDeclarationTrustStore() override { return &mCdKeysTrustStore; }
 
-    void SetRevocationDelegate(DeviceAttestationRevocationDelegate * revocationDelegate)
+    CHIP_ERROR TrySettingRevocationDelegate(DeviceAttestationRevocationDelegate * revocationDelegate) override
```
NIT: Why its renamed from `SetRevocationDelegate` to `TrySettingRevocationDelegate`? I do not see any convention in the codebase with Try* naming for setters. The convention is that setter might fail if it returns CHIP_ERROR. "Try" suggests that something might fail temporally, so caller should repeat call attempt after some time. It's not the case here...

shubhamdp replied:
It was the suggestion from @tehampson in https://github.com/project-chip/connectedhomeip/pull/37134#discussion_r1932304885.

Wasn't aware about the `Try*` convention. Agree, we can keep `SetRevocationDelegate()`.

arkq replied:
I'm not sure whether we have any formal convention for that but that's how I read it in common APIs like POSIX or glib.

shubhamdp committed:
rename TrySettingRevocationDelegate to SetRevocationDelegate
